language/stdlib/modules/Libra.move

```diff
@@ -1158,73 +1126,177 @@ module Libra {
     /// aggregate value exceeds the CoinInfo.total_value.  However, that property involves summations over
     /// all resources and is beyond the capabilities of the specification logic or the prover, currently.
 
-    spec module {
-        /// The permission "MintCurrency(type)" is granted to TreasuryCompliance [B11].
-        apply Roles::AbortsIfNotTreasuryCompliance{account: tc_account} to register_SCS_currency<CoinType>;
-
-        /// The permission "BurnCurrency(type)" is granted to TreasuryCompliance [B12].
-        apply Roles::AbortsIfNotTreasuryCompliance{account: tc_account} to register_SCS_currency<CoinType>;
 
-        /// The permission "PreburnCurrency(type)" is granted to DesignatedDealer [B13].
-        apply Roles::AbortsIfNotDesignatedDealer to publish_preburn_to_account<CoinType>;
-
-        /// The permission "UpdateExchangeRate(type)" is granted to TreasuryCompliance [B14].
-        apply Roles::AbortsIfNotTreasuryCompliance{account: tc_account} to update_lbr_exchange_rate<FromCoinType>;
-    }
+    /// ## Minting
 
     spec schema TotalValueNotIncrease<CoinType> {
         /// The total amount of currency does not increase.
         ensures old(spec_is_currency<CoinType>())
             ==> spec_currency_info<CoinType>().total_value <= old(spec_currency_info<CoinType>().total_value);
     }
+    spec schema PreserveMintCapExistence<CoinType> {
+        /// The existence of MintCapability is preserved.
+        ensures forall addr: address:
+            old(exists<MintCapability<CoinType>>(addr)) ==>
+                exists<MintCapability<CoinType>>(addr);
+    }
+    spec schema PreserveMintCapAbsence<CoinType> {
+        /// The absence of MintCapability is preserved.
+        ensures forall addr: address:
+            old(!exists<MintCapability<CoinType>>(addr)) ==>
+                !exists<MintCapability<CoinType>>(addr);
+    }
     spec module {
         /// Only mint functions can increase the total amount of currency [B11].
         apply TotalValueNotIncrease<CoinType> to *<CoinType>
             except mint<CoinType>, mint_with_capability<CoinType>;
+
+        /// In order to successfully call `mint` and `mint_with_capability`, MintCapability is
+        /// required. MintCapability must be only granted to a TreasuryCompliance account [B11].
+        /// Only `register_SCS_currency` creates MintCapability, which must abort if the account
+        /// does not have the TreasuryCompliance role [B17].
+        apply PreserveMintCapAbsence<CoinType> to *<CoinType> except register_SCS_currency<CoinType>;
```

Comment:
I think there's a soundness issue here, unless something has changed recently.  Suppose some other module defines a function that calls register_currency, gets a capability, and publishes it somewhere.  Then, this spec could be violated by a function that PreserveMintAbsence is not applied to.

With global invariants, we can put all such modules on the command line (Coin1, Coin2, LBR? in this case).  But I don't think that works with "apply".

Reply:
A lot of these schema-based postconditions seem to be specifications which are (more nicely and efficiently, in my opinion) expressed using two-state invariants (specified using the attribute "update").  The guard in the two-state invariant would contain the access-control predicate saying something like "signer has this or that role".  Is there a reason why we prefer these apply-style postconditions rather than two-state invariants?  Could it be the case that "signer has this/that role" is not expressible in our spec language?  

@wrwg, @sblackshear .

Reply:
> Could it be the case that "signer has this/that role" is not expressible in our spec language?
@shazqadeer, yes, `account: &signer` is a function parameter, and I believe that our spec lang currently does not support using function parameters in invariants. Also, some Move functions does not have such an "account" parameter, and some other Move functions have one with different names (e.g., lr_account, tc_account).

Reply:
I share David and Shaz's concerns about this `apply` + `Preserve`-based specification methodology. I don't think this is expressing the property of interest.

Reply:
I added a TODO for this in the code. It seems that we don't have a quick solution for this currently.